{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from math import sqrt\n",
    "import copy\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>53</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "      <th>56</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>59</th>\n",
       "      <th>60</th>\n",
       "      <th>61</th>\n",
       "      <th>62</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Telomere_attrition</th>\n",
       "      <td>ATM</td>\n",
       "      <td>FEN1</td>\n",
       "      <td>SIRT1</td>\n",
       "      <td>TERC</td>\n",
       "      <td>TERF1</td>\n",
       "      <td>TERF2</td>\n",
       "      <td>TERT</td>\n",
       "      <td>XRCC6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stem_cell_exhaustion</th>\n",
       "      <td>FOXM1</td>\n",
       "      <td>MAPK14</td>\n",
       "      <td>PLAU</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>STAT5A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mitochondrial_dysfunction</th>\n",
       "      <td>APTX</td>\n",
       "      <td>ATP5O</td>\n",
       "      <td>CAT</td>\n",
       "      <td>COQ7</td>\n",
       "      <td>HSPD1</td>\n",
       "      <td>IGF1R</td>\n",
       "      <td>MT-CO1</td>\n",
       "      <td>POLG</td>\n",
       "      <td>RAD51</td>\n",
       "      <td>RNTL</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genomic_instability</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>APEX1</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>BLM</td>\n",
       "      <td>BMI1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>BUB1B</td>\n",
       "      <td>CCNA2</td>\n",
       "      <td>CDK1</td>\n",
       "      <td>...</td>\n",
       "      <td>TOP1</td>\n",
       "      <td>TOP2A</td>\n",
       "      <td>TOP2B</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TR</td>\n",
       "      <td>UCHL1</td>\n",
       "      <td>WRN</td>\n",
       "      <td>XPA</td>\n",
       "      <td>XRCC5</td>\n",
       "      <td>XRCC6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epigenetic_alterations</th>\n",
       "      <td>CDK1</td>\n",
       "      <td>CREB1</td>\n",
       "      <td>DDIT3</td>\n",
       "      <td>HBP1</td>\n",
       "      <td>HDAC1</td>\n",
       "      <td>HDAC2</td>\n",
       "      <td>HMGB1</td>\n",
       "      <td>HMGB2</td>\n",
       "      <td>PML</td>\n",
       "      <td>PROP1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Deregulated_nutrient_sensing</th>\n",
       "      <td>AGPAT2</td>\n",
       "      <td>AKT1</td>\n",
       "      <td>APOE</td>\n",
       "      <td>AR</td>\n",
       "      <td>BSCL2</td>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>DCY5</td>\n",
       "      <td>FGFR1</td>\n",
       "      <td>FOXO1</td>\n",
       "      <td>GH1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cellular_senescence</th>\n",
       "      <td>ATF2</td>\n",
       "      <td>BAX</td>\n",
       "      <td>BMI1</td>\n",
       "      <td>BUB3</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>CTGF</td>\n",
       "      <td>E2F1</td>\n",
       "      <td>EGF</td>\n",
       "      <td>EP300</td>\n",
       "      <td>EPOR</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Altered_intercellular_communication</th>\n",
       "      <td>AIFM1</td>\n",
       "      <td>APOC3</td>\n",
       "      <td>AR</td>\n",
       "      <td>BDNF</td>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>CEBPA</td>\n",
       "      <td>CEBPB</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGR1</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Loss_of_proteostasis</th>\n",
       "      <td>A2M</td>\n",
       "      <td>APP</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>BSCL2</td>\n",
       "      <td>DDIT3</td>\n",
       "      <td>EEF2</td>\n",
       "      <td>GSTA4</td>\n",
       "      <td>HSPA9</td>\n",
       "      <td>MAPT</td>\n",
       "      <td>MSRA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         0       1      2      3        4   \\\n",
       "Telomere_attrition                      ATM    FEN1  SIRT1   TERC    TERF1   \n",
       "Stem_cell_exhaustion                  FOXM1  MAPK14   PLAU  STAT3   STAT5A   \n",
       "Mitochondrial_dysfunction              APTX   ATP5O    CAT   COQ7    HSPD1   \n",
       "Genomic_instability                    ABL1   APEX1   BAK1    BLM     BMI1   \n",
       "Epigenetic_alterations                 CDK1   CREB1  DDIT3   HBP1    HDAC1   \n",
       "Deregulated_nutrient_sensing         AGPAT2    AKT1   APOE     AR    BSCL2   \n",
       "Cellular_senescence                    ATF2     BAX   BMI1   BUB3   CREBBP   \n",
       "Altered_intercellular_communication   AIFM1   APOC3     AR   BDNF  CACNA1A   \n",
       "Loss_of_proteostasis                    A2M     APP   BCL2  BSCL2    DDIT3   \n",
       "\n",
       "                                          5       6      7      8      9   \\\n",
       "Telomere_attrition                     TERF2    TERT  XRCC6    NaN    NaN   \n",
       "Stem_cell_exhaustion                     NaN     NaN    NaN    NaN    NaN   \n",
       "Mitochondrial_dysfunction              IGF1R  MT-CO1   POLG  RAD51   RNTL   \n",
       "Genomic_instability                    BRCA1   BRCA2  BUB1B  CCNA2   CDK1   \n",
       "Epigenetic_alterations                 HDAC2   HMGB1  HMGB2    PML  PROP1   \n",
       "Deregulated_nutrient_sensing         CACNA1A    DCY5  FGFR1  FOXO1    GH1   \n",
       "Cellular_senescence                     CTGF    E2F1    EGF  EP300   EPOR   \n",
       "Altered_intercellular_communication    CEBPA   CEBPB   EGFR   EGR1  ERBB2   \n",
       "Loss_of_proteostasis                    EEF2   GSTA4  HSPA9   MAPT   MSRA   \n",
       "\n",
       "                                     ...    53     54     55    56   57  \\\n",
       "Telomere_attrition                   ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Stem_cell_exhaustion                 ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Mitochondrial_dysfunction            ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Genomic_instability                  ...  TOP1  TOP2A  TOP2B  TP53   TR   \n",
       "Epigenetic_alterations               ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Deregulated_nutrient_sensing         ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Cellular_senescence                  ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Altered_intercellular_communication  ...   NaN    NaN    NaN   NaN  NaN   \n",
       "Loss_of_proteostasis                 ...   NaN    NaN    NaN   NaN  NaN   \n",
       "\n",
       "                                        58   59   60     61     62  \n",
       "Telomere_attrition                     NaN  NaN  NaN    NaN    NaN  \n",
       "Stem_cell_exhaustion                   NaN  NaN  NaN    NaN    NaN  \n",
       "Mitochondrial_dysfunction              NaN  NaN  NaN    NaN    NaN  \n",
       "Genomic_instability                  UCHL1  WRN  XPA  XRCC5  XRCC6  \n",
       "Epigenetic_alterations                 NaN  NaN  NaN    NaN    NaN  \n",
       "Deregulated_nutrient_sensing           NaN  NaN  NaN    NaN    NaN  \n",
       "Cellular_senescence                    NaN  NaN  NaN    NaN    NaN  \n",
       "Altered_intercellular_communication    NaN  NaN  NaN    NaN    NaN  \n",
       "Loss_of_proteostasis                   NaN  NaN  NaN    NaN    NaN  \n",
       "\n",
       "[9 rows x 63 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proto_gmt = pd.read_csv(\"data_for_gmt.csv\")\n",
    "gmt = proto_gmt.T\n",
    "gmt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TPT1</td>\n",
       "      <td>14.487080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPL41</td>\n",
       "      <td>14.448307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HLA-A</td>\n",
       "      <td>14.415246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TMSB4X</td>\n",
       "      <td>14.405954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UBA52</td>\n",
       "      <td>14.400992</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene  Expression\n",
       "0    TPT1   14.487080\n",
       "1   RPL41   14.448307\n",
       "2   HLA-A   14.415246\n",
       "3  TMSB4X   14.405954\n",
       "4   UBA52   14.400992"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression = pd.read_csv(\"gene_expression_complete.csv\", names = ['Gene', 'Expression'], skiprows = [0])\n",
    "expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have a gmt file and a list of expression values, let's confirm that we can map from one to the other. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "So, 58 of \n",
      "63 genes have mapped.\n"
     ]
    }
   ],
   "source": [
    "count = 0 \n",
    "for i in gmt.iloc[3,:]:\n",
    "    if i in set(expression.Gene):\n",
    "        count += 1\n",
    "        \n",
    "print('So, ' + str(count) + ' of ')\n",
    "# This is inelegant, as there are some NaN values, etc, but it confirms that our labels map.    \n",
    "print(str(len(gmt.iloc[3,:])) + ' genes have mapped.')   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                     Enrichment_Score\n",
      "Telomere_attrition                             142.03\n",
      "Stem_cell_exhaustion                            86.22\n",
      "Mitochondrial_dysfunction                      222.73\n",
      "Genomic_instability                            243.03\n",
      "Epigenetic_alterations                         120.60\n",
      "Deregulated_nutrient_sensing                   130.65\n",
      "Cellular_senescence                            206.86\n",
      "Altered_intercellular_communication            188.48\n",
      "Loss_of_proteostasis                           257.67\n"
     ]
    }
   ],
   "source": [
    "# Now we will try to do GSEA. Note that our 'expression' file is already ranked -- \n",
    "# it is a list of genes with their expression in treated relative to control, \n",
    "# ranked from highest to lowest expression\n",
    "\n",
    "def calculate_enrichment_scores(expression, gmt):\n",
    "    '''\n",
    "    a function that calculates the enrichment score of a gene set in your expression data\n",
    "    inputs: \n",
    "        expression data should be a list of gene ids, ordered by relative fold change between \n",
    "        treatment and control such that the gene with the greatest fold-change appears at the \n",
    "        top of the list and vice versa\n",
    "        \n",
    "        gmt_info should be your gmt file, parsed into a pandas DataFrame\n",
    "        \n",
    "    outputs:\n",
    "        results, a DataFrame of gene set names and enrichment scores\n",
    "        \n",
    "    '''\n",
    "    Enrichment_Score_dict = {}\n",
    "\n",
    "    # for row in gmt\n",
    "    for row in gmt.index:\n",
    "\n",
    "        # find the geneset of interest; it is the intersection of the genes from the GMT file and the genes in the expression dataset\n",
    "        full_gene_set = {x for x in gmt.loc[row,:] if pd.notna(x)}\n",
    "        expression_set = set(expression.Gene) \n",
    "        gene_set = expression_set.intersection(full_gene_set)\n",
    "\n",
    "        if len(gene_set) == 0:\n",
    "            Enrichment_Score_dict[row] = 0\n",
    "            print(str(row) + ' was 0')\n",
    "\n",
    "        else: \n",
    "            # Now to set up our random walk penalties\n",
    "            reward_amount = sqrt((len(expression) - len(gene_set))/(len(gene_set)))\n",
    "            penalty_amount = -1 * sqrt(len(gene_set) / (len(expression) - len(gene_set)))\n",
    "            # let's take a random walk\n",
    "            expression = expression.assign(reward_or_penalty = [reward_amount if x in gene_set else penalty_amount for x in expression.Gene])\n",
    "            expression = expression.assign(the_walk = np.cumsum(expression.reward_or_penalty))\n",
    "\n",
    "            # now let's check that the random walk ends with a zero (as it should)\n",
    "            if int(expression.the_walk.tail(1)) != 0:\n",
    "                print(int(expression.the_walk.tail(1)))\n",
    "                raise ValueError(\"your brownian bridge is broken\")  ## just checking\n",
    "\n",
    "            Enrichment_Score_dict[row] = round(max(expression.the_walk),2)\n",
    "\n",
    "    results = pd.DataFrame.from_dict(Enrichment_Score_dict, orient = 'index', columns = ['Enrichment_Score'])\n",
    "    return(results)\n",
    "\n",
    "results = calculate_enrichment_scores(expression, gmt)    \n",
    "print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we need to find a critical threshold value to decide what constitutes statistically significant enrichment.\n",
    "We will do this empirically by generating random data and observing the fraction of times we obtain a value greater than our true values by chance alone. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def emperic_p_values(expression, gmt, res, n_iterations):\n",
    "\n",
    "    distribution = copy.copy(res)\n",
    "    for i in range(n_iterations):\n",
    "        pretend_ranked_list = pd.DataFrame(expression.Gene.sample(frac = 1))\n",
    "        rando_scores = calculate_enrichment_scores(pretend_ranked_list, gmt)\n",
    "        distribution = pd.concat([distribution, rando_scores], axis = 1)\n",
    "    \n",
    "    raw_p = {}\n",
    "    corrected_p = {}\n",
    "\n",
    "    # take each set of enrichment scores - the first is our result, the others occured by chance alone\n",
    "    for k in distribution.index:\n",
    "        list_of_values = list(distribution.loc[k,:])\n",
    "        number_greater = 0\n",
    "        for l in list_of_values[1:]:\n",
    "            if l >= list_of_values[0]:\n",
    "                number_greater += 1\n",
    "        # the raw p value is the fraction of randomly generated enrichment scores that were equal to or greater than our result        \n",
    "        raw_p[k] = (float(number_greater)/float(len(list_of_values[1:])))\n",
    "\n",
    "    p_values = pd.DataFrame.from_dict(raw_p, orient = 'index', columns = ['Raw_p_value'])\n",
    "\n",
    "    # now we can correct for multiple testing by determining an appropriate p value threshold with Bonferroni correction\n",
    "    p_values = p_values.assign(p_value_threshold = 0.05 / distribution.shape[0])\n",
    "    p_values = p_values.assign(Significantly_enriched = p_values.Raw_p_value <= p_values.p_value_threshold)\n",
    "    return(p_values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Raw_p_value</th>\n",
       "      <th>p_value_threshold</th>\n",
       "      <th>Significantly_enriched</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Telomere_attrition</th>\n",
       "      <td>0.064</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stem_cell_exhaustion</th>\n",
       "      <td>0.348</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mitochondrial_dysfunction</th>\n",
       "      <td>0.002</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genomic_instability</th>\n",
       "      <td>0.002</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epigenetic_alterations</th>\n",
       "      <td>0.155</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Deregulated_nutrient_sensing</th>\n",
       "      <td>0.118</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cellular_senescence</th>\n",
       "      <td>0.003</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Altered_intercellular_communication</th>\n",
       "      <td>0.007</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Loss_of_proteostasis</th>\n",
       "      <td>0.001</td>\n",
       "      <td>0.005556</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Raw_p_value  p_value_threshold  \\\n",
       "Telomere_attrition                         0.064           0.005556   \n",
       "Stem_cell_exhaustion                       0.348           0.005556   \n",
       "Mitochondrial_dysfunction                  0.002           0.005556   \n",
       "Genomic_instability                        0.002           0.005556   \n",
       "Epigenetic_alterations                     0.155           0.005556   \n",
       "Deregulated_nutrient_sensing               0.118           0.005556   \n",
       "Cellular_senescence                        0.003           0.005556   \n",
       "Altered_intercellular_communication        0.007           0.005556   \n",
       "Loss_of_proteostasis                       0.001           0.005556   \n",
       "\n",
       "                                     Significantly_enriched  \n",
       "Telomere_attrition                                    False  \n",
       "Stem_cell_exhaustion                                  False  \n",
       "Mitochondrial_dysfunction                              True  \n",
       "Genomic_instability                                    True  \n",
       "Epigenetic_alterations                                False  \n",
       "Deregulated_nutrient_sensing                          False  \n",
       "Cellular_senescence                                    True  \n",
       "Altered_intercellular_communication                   False  \n",
       "Loss_of_proteostasis                                   True  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "emperic_p_values(expression, gmt,results, 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, as a comparison, let's perform GSEA using canonical KEGG pathways. I will need a bit of code to parse the KEGG file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_gmt_dict = {}\n",
    "with open('c2.cp.kegg.v6.2.symbols.gmt') as f:\n",
    "    for line in f:\n",
    "        line_bits = line.split()\n",
    "        kegg_gmt_dict[line_bits[0]] = line_bits[1:]\n",
    "\n",
    "kegg_gmt_frame = pd.DataFrame.from_dict(kegg_gmt_dict, orient='index' )\n",
    "kegg_gmt = kegg_gmt_frame.iloc[:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Enrichment_Score\n",
      "KEGG_GLYCOLYSIS_GLUCONEOGENESIS                               210.27\n",
      "KEGG_CITRATE_CYCLE_TCA_CYCLE                                  334.74\n",
      "KEGG_PENTOSE_PHOSPHATE_PATHWAY                                230.16\n",
      "KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS                  33.88\n",
      "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM                          240.10\n",
      "KEGG_GALACTOSE_METABOLISM                                     198.41\n",
      "KEGG_ASCORBATE_AND_ALDARATE_METABOLISM                         28.75\n",
      "KEGG_FATTY_ACID_METABOLISM                                    172.55\n",
      "KEGG_STEROID_BIOSYNTHESIS                                     114.66\n",
      "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS                            47.12\n",
      "KEGG_STEROID_HORMONE_BIOSYNTHESIS                              33.05\n",
      "KEGG_OXIDATIVE_PHOSPHORYLATION                                572.66\n",
      "KEGG_PURINE_METABOLISM                                        225.94\n",
      "KEGG_PYRIMIDINE_METABOLISM                                    328.08\n",
      "KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM                95.56\n",
      "KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM                    1.75\n",
      "KEGG_CYSTEINE_AND_METHIONINE_METABOLISM                       132.27\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION                254.73\n",
      "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS               149.71\n",
      "KEGG_LYSINE_DEGRADATION                                       174.10\n",
      "KEGG_ARGININE_AND_PROLINE_METABOLISM                           84.82\n",
      "KEGG_HISTIDINE_METABOLISM                                      62.50\n",
      "KEGG_TYROSINE_METABOLISM                                       47.45\n",
      "KEGG_PHENYLALANINE_METABOLISM                                  41.55\n",
      "KEGG_TRYPTOPHAN_METABOLISM                                    113.15\n",
      "KEGG_BETA_ALANINE_METABOLISM                                  102.75\n",
      "KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM                        29.50\n",
      "KEGG_SELENOAMINO_ACID_METABOLISM                              149.81\n",
      "KEGG_GLUTATHIONE_METABOLISM                                   210.19\n",
      "KEGG_STARCH_AND_SUCROSE_METABOLISM                             71.15\n",
      "...                                                              ...\n",
      "KEGG_PRION_DISEASES                                           200.40\n",
      "KEGG_VIBRIO_CHOLERAE_INFECTION                                284.10\n",
      "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_...            289.65\n",
      "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION                    404.20\n",
      "KEGG_LEISHMANIA_INFECTION                                     406.42\n",
      "KEGG_PATHWAYS_IN_CANCER                                       289.80\n",
      "KEGG_COLORECTAL_CANCER                                        269.03\n",
      "KEGG_RENAL_CELL_CARCINOMA                                     267.78\n",
      "KEGG_PANCREATIC_CANCER                                        286.59\n",
      "KEGG_ENDOMETRIAL_CANCER                                       239.65\n",
      "KEGG_GLIOMA                                                   220.93\n",
      "KEGG_PROSTATE_CANCER                                          267.54\n",
      "KEGG_THYROID_CANCER                                           188.23\n",
      "KEGG_BASAL_CELL_CARCINOMA                                      10.79\n",
      "KEGG_MELANOMA                                                  67.07\n",
      "KEGG_BLADDER_CANCER                                           135.74\n",
      "KEGG_CHRONIC_MYELOID_LEUKEMIA                                 351.69\n",
      "KEGG_ACUTE_MYELOID_LEUKEMIA                                   308.81\n",
      "KEGG_SMALL_CELL_LUNG_CANCER                                   192.70\n",
      "KEGG_NON_SMALL_CELL_LUNG_CANCER                               252.78\n",
      "KEGG_ASTHMA                                                   188.65\n",
      "KEGG_AUTOIMMUNE_THYROID_DISEASE                               217.89\n",
      "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS                             113.08\n",
      "KEGG_ALLOGRAFT_REJECTION                                      265.35\n",
      "KEGG_GRAFT_VERSUS_HOST_DISEASE                                304.81\n",
      "KEGG_PRIMARY_IMMUNODEFICIENCY                                 254.20\n",
      "KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM                           45.50\n",
      "KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYO...             83.80\n",
      "KEGG_DILATED_CARDIOMYOPATHY                                    46.55\n",
      "KEGG_VIRAL_MYOCARDITIS                                        285.89\n",
      "\n",
      "[186 rows x 1 columns]\n"
     ]
    }
   ],
   "source": [
    "kegg_results = calculate_enrichment_scores(expression, kegg_gmt)    \n",
    "print(kegg_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Raw_p_value</th>\n",
       "      <th>p_value_threshold</th>\n",
       "      <th>Significantly_enriched</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG_GLYCOLYSIS_GLUCONEOGENESIS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CITRATE_CYCLE_TCA_CYCLE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PENTOSE_PHOSPHATE_PATHWAY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GALACTOSE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_OXIDATIVE_PHOSPHORYLATION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PURINE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PYRIMIDINE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GLUTATHIONE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_N_GLYCAN_BIOSYNTHESIS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_OTHER_GLYCAN_DEGRADATION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_INOSITOL_PHOSPHATE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PROPANOATE_METABOLISM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_AMINOACYL_TRNA_BIOSYNTHESIS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_RIBOSOME</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_RNA_DEGRADATION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_RNA_POLYMERASE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_BASAL_TRANSCRIPTION_FACTORS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_DNA_REPLICATION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_SPLICEOSOME</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PROTEASOME</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PROTEIN_EXPORT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_BASE_EXCISION_REPAIR</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_NUCLEOTIDE_EXCISION_REPAIR</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_MISMATCH_REPAIR</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_MAPK_SIGNALING_PATHWAY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_LONG_TERM_POTENTIATION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_NEUROTROPHIN_SIGNALING_PATHWAY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_REGULATION_OF_ACTIN_CYTOSKELETON</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_INSULIN_SIGNALING_PATHWAY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_TYPE_I_DIABETES_MELLITUS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ALZHEIMERS_DISEASE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PARKINSONS_DISEASE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_HUNTINGTONS_DISEASE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PRION_DISEASES</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_VIBRIO_CHOLERAE_INFECTION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_LEISHMANIA_INFECTION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PATHWAYS_IN_CANCER</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_COLORECTAL_CANCER</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_RENAL_CELL_CARCINOMA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PANCREATIC_CANCER</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ENDOMETRIAL_CANCER</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GLIOMA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PROSTATE_CANCER</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CHRONIC_MYELOID_LEUKEMIA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ACUTE_MYELOID_LEUKEMIA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_NON_SMALL_CELL_LUNG_CANCER</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_AUTOIMMUNE_THYROID_DISEASE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ALLOGRAFT_REJECTION</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GRAFT_VERSUS_HOST_DISEASE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PRIMARY_IMMUNODEFICIENCY</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_VIRAL_MYOCARDITIS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    Raw_p_value  \\\n",
       "KEGG_GLYCOLYSIS_GLUCONEOGENESIS                             0.0   \n",
       "KEGG_CITRATE_CYCLE_TCA_CYCLE                                0.0   \n",
       "KEGG_PENTOSE_PHOSPHATE_PATHWAY                              0.0   \n",
       "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM                        0.0   \n",
       "KEGG_GALACTOSE_METABOLISM                                   0.0   \n",
       "KEGG_OXIDATIVE_PHOSPHORYLATION                              0.0   \n",
       "KEGG_PURINE_METABOLISM                                      0.0   \n",
       "KEGG_PYRIMIDINE_METABOLISM                                  0.0   \n",
       "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION              0.0   \n",
       "KEGG_GLUTATHIONE_METABOLISM                                 0.0   \n",
       "KEGG_N_GLYCAN_BIOSYNTHESIS                                  0.0   \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                               0.0   \n",
       "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM            0.0   \n",
       "KEGG_INOSITOL_PHOSPHATE_METABOLISM                          0.0   \n",
       "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BI...          0.0   \n",
       "KEGG_PROPANOATE_METABOLISM                                  0.0   \n",
       "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS                        0.0   \n",
       "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS                            0.0   \n",
       "KEGG_RIBOSOME                                               0.0   \n",
       "KEGG_RNA_DEGRADATION                                        0.0   \n",
       "KEGG_RNA_POLYMERASE                                         0.0   \n",
       "KEGG_BASAL_TRANSCRIPTION_FACTORS                            0.0   \n",
       "KEGG_DNA_REPLICATION                                        0.0   \n",
       "KEGG_SPLICEOSOME                                            0.0   \n",
       "KEGG_PROTEASOME                                             0.0   \n",
       "KEGG_PROTEIN_EXPORT                                         0.0   \n",
       "KEGG_BASE_EXCISION_REPAIR                                   0.0   \n",
       "KEGG_NUCLEOTIDE_EXCISION_REPAIR                             0.0   \n",
       "KEGG_MISMATCH_REPAIR                                        0.0   \n",
       "KEGG_MAPK_SIGNALING_PATHWAY                                 0.0   \n",
       "...                                                         ...   \n",
       "KEGG_LONG_TERM_POTENTIATION                                 0.0   \n",
       "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY                         0.0   \n",
       "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON                       0.0   \n",
       "KEGG_INSULIN_SIGNALING_PATHWAY                              0.0   \n",
       "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY                        0.0   \n",
       "KEGG_TYPE_I_DIABETES_MELLITUS                               0.0   \n",
       "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION               0.0   \n",
       "KEGG_ALZHEIMERS_DISEASE                                     0.0   \n",
       "KEGG_PARKINSONS_DISEASE                                     0.0   \n",
       "KEGG_HUNTINGTONS_DISEASE                                    0.0   \n",
       "KEGG_PRION_DISEASES                                         0.0   \n",
       "KEGG_VIBRIO_CHOLERAE_INFECTION                              0.0   \n",
       "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_...          0.0   \n",
       "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION                  0.0   \n",
       "KEGG_LEISHMANIA_INFECTION                                   0.0   \n",
       "KEGG_PATHWAYS_IN_CANCER                                     0.0   \n",
       "KEGG_COLORECTAL_CANCER                                      0.0   \n",
       "KEGG_RENAL_CELL_CARCINOMA                                   0.0   \n",
       "KEGG_PANCREATIC_CANCER                                      0.0   \n",
       "KEGG_ENDOMETRIAL_CANCER                                     0.0   \n",
       "KEGG_GLIOMA                                                 0.0   \n",
       "KEGG_PROSTATE_CANCER                                        0.0   \n",
       "KEGG_CHRONIC_MYELOID_LEUKEMIA                               0.0   \n",
       "KEGG_ACUTE_MYELOID_LEUKEMIA                                 0.0   \n",
       "KEGG_NON_SMALL_CELL_LUNG_CANCER                             0.0   \n",
       "KEGG_AUTOIMMUNE_THYROID_DISEASE                             0.0   \n",
       "KEGG_ALLOGRAFT_REJECTION                                    0.0   \n",
       "KEGG_GRAFT_VERSUS_HOST_DISEASE                              0.0   \n",
       "KEGG_PRIMARY_IMMUNODEFICIENCY                               0.0   \n",
       "KEGG_VIRAL_MYOCARDITIS                                      0.0   \n",
       "\n",
       "                                                    p_value_threshold  \\\n",
       "KEGG_GLYCOLYSIS_GLUCONEOGENESIS                              0.000269   \n",
       "KEGG_CITRATE_CYCLE_TCA_CYCLE                                 0.000269   \n",
       "KEGG_PENTOSE_PHOSPHATE_PATHWAY                               0.000269   \n",
       "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM                         0.000269   \n",
       "KEGG_GALACTOSE_METABOLISM                                    0.000269   \n",
       "KEGG_OXIDATIVE_PHOSPHORYLATION                               0.000269   \n",
       "KEGG_PURINE_METABOLISM                                       0.000269   \n",
       "KEGG_PYRIMIDINE_METABOLISM                                   0.000269   \n",
       "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION               0.000269   \n",
       "KEGG_GLUTATHIONE_METABOLISM                                  0.000269   \n",
       "KEGG_N_GLYCAN_BIOSYNTHESIS                                   0.000269   \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                                0.000269   \n",
       "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM             0.000269   \n",
       "KEGG_INOSITOL_PHOSPHATE_METABOLISM                           0.000269   \n",
       "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BI...           0.000269   \n",
       "KEGG_PROPANOATE_METABOLISM                                   0.000269   \n",
       "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS                         0.000269   \n",
       "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS                             0.000269   \n",
       "KEGG_RIBOSOME                                                0.000269   \n",
       "KEGG_RNA_DEGRADATION                                         0.000269   \n",
       "KEGG_RNA_POLYMERASE                                          0.000269   \n",
       "KEGG_BASAL_TRANSCRIPTION_FACTORS                             0.000269   \n",
       "KEGG_DNA_REPLICATION                                         0.000269   \n",
       "KEGG_SPLICEOSOME                                             0.000269   \n",
       "KEGG_PROTEASOME                                              0.000269   \n",
       "KEGG_PROTEIN_EXPORT                                          0.000269   \n",
       "KEGG_BASE_EXCISION_REPAIR                                    0.000269   \n",
       "KEGG_NUCLEOTIDE_EXCISION_REPAIR                              0.000269   \n",
       "KEGG_MISMATCH_REPAIR                                         0.000269   \n",
       "KEGG_MAPK_SIGNALING_PATHWAY                                  0.000269   \n",
       "...                                                               ...   \n",
       "KEGG_LONG_TERM_POTENTIATION                                  0.000269   \n",
       "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY                          0.000269   \n",
       "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON                        0.000269   \n",
       "KEGG_INSULIN_SIGNALING_PATHWAY                               0.000269   \n",
       "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY                         0.000269   \n",
       "KEGG_TYPE_I_DIABETES_MELLITUS                                0.000269   \n",
       "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION                0.000269   \n",
       "KEGG_ALZHEIMERS_DISEASE                                      0.000269   \n",
       "KEGG_PARKINSONS_DISEASE                                      0.000269   \n",
       "KEGG_HUNTINGTONS_DISEASE                                     0.000269   \n",
       "KEGG_PRION_DISEASES                                          0.000269   \n",
       "KEGG_VIBRIO_CHOLERAE_INFECTION                               0.000269   \n",
       "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_...           0.000269   \n",
       "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION                   0.000269   \n",
       "KEGG_LEISHMANIA_INFECTION                                    0.000269   \n",
       "KEGG_PATHWAYS_IN_CANCER                                      0.000269   \n",
       "KEGG_COLORECTAL_CANCER                                       0.000269   \n",
       "KEGG_RENAL_CELL_CARCINOMA                                    0.000269   \n",
       "KEGG_PANCREATIC_CANCER                                       0.000269   \n",
       "KEGG_ENDOMETRIAL_CANCER                                      0.000269   \n",
       "KEGG_GLIOMA                                                  0.000269   \n",
       "KEGG_PROSTATE_CANCER                                         0.000269   \n",
       "KEGG_CHRONIC_MYELOID_LEUKEMIA                                0.000269   \n",
       "KEGG_ACUTE_MYELOID_LEUKEMIA                                  0.000269   \n",
       "KEGG_NON_SMALL_CELL_LUNG_CANCER                              0.000269   \n",
       "KEGG_AUTOIMMUNE_THYROID_DISEASE                              0.000269   \n",
       "KEGG_ALLOGRAFT_REJECTION                                     0.000269   \n",
       "KEGG_GRAFT_VERSUS_HOST_DISEASE                               0.000269   \n",
       "KEGG_PRIMARY_IMMUNODEFICIENCY                                0.000269   \n",
       "KEGG_VIRAL_MYOCARDITIS                                       0.000269   \n",
       "\n",
       "                                                    Significantly_enriched  \n",
       "KEGG_GLYCOLYSIS_GLUCONEOGENESIS                                       True  \n",
       "KEGG_CITRATE_CYCLE_TCA_CYCLE                                          True  \n",
       "KEGG_PENTOSE_PHOSPHATE_PATHWAY                                        True  \n",
       "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM                                  True  \n",
       "KEGG_GALACTOSE_METABOLISM                                             True  \n",
       "KEGG_OXIDATIVE_PHOSPHORYLATION                                        True  \n",
       "KEGG_PURINE_METABOLISM                                                True  \n",
       "KEGG_PYRIMIDINE_METABOLISM                                            True  \n",
       "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION                        True  \n",
       "KEGG_GLUTATHIONE_METABOLISM                                           True  \n",
       "KEGG_N_GLYCAN_BIOSYNTHESIS                                            True  \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                                         True  \n",
       "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM                      True  \n",
       "KEGG_INOSITOL_PHOSPHATE_METABOLISM                                    True  \n",
       "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BI...                    True  \n",
       "KEGG_PROPANOATE_METABOLISM                                            True  \n",
       "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS                                  True  \n",
       "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS                                      True  \n",
       "KEGG_RIBOSOME                                                         True  \n",
       "KEGG_RNA_DEGRADATION                                                  True  \n",
       "KEGG_RNA_POLYMERASE                                                   True  \n",
       "KEGG_BASAL_TRANSCRIPTION_FACTORS                                      True  \n",
       "KEGG_DNA_REPLICATION                                                  True  \n",
       "KEGG_SPLICEOSOME                                                      True  \n",
       "KEGG_PROTEASOME                                                       True  \n",
       "KEGG_PROTEIN_EXPORT                                                   True  \n",
       "KEGG_BASE_EXCISION_REPAIR                                             True  \n",
       "KEGG_NUCLEOTIDE_EXCISION_REPAIR                                       True  \n",
       "KEGG_MISMATCH_REPAIR                                                  True  \n",
       "KEGG_MAPK_SIGNALING_PATHWAY                                           True  \n",
       "...                                                                    ...  \n",
       "KEGG_LONG_TERM_POTENTIATION                                           True  \n",
       "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY                                   True  \n",
       "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON                                 True  \n",
       "KEGG_INSULIN_SIGNALING_PATHWAY                                        True  \n",
       "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY                                  True  \n",
       "KEGG_TYPE_I_DIABETES_MELLITUS                                         True  \n",
       "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION                         True  \n",
       "KEGG_ALZHEIMERS_DISEASE                                               True  \n",
       "KEGG_PARKINSONS_DISEASE                                               True  \n",
       "KEGG_HUNTINGTONS_DISEASE                                              True  \n",
       "KEGG_PRION_DISEASES                                                   True  \n",
       "KEGG_VIBRIO_CHOLERAE_INFECTION                                        True  \n",
       "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_...                    True  \n",
       "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION                            True  \n",
       "KEGG_LEISHMANIA_INFECTION                                             True  \n",
       "KEGG_PATHWAYS_IN_CANCER                                               True  \n",
       "KEGG_COLORECTAL_CANCER                                                True  \n",
       "KEGG_RENAL_CELL_CARCINOMA                                             True  \n",
       "KEGG_PANCREATIC_CANCER                                                True  \n",
       "KEGG_ENDOMETRIAL_CANCER                                               True  \n",
       "KEGG_GLIOMA                                                           True  \n",
       "KEGG_PROSTATE_CANCER                                                  True  \n",
       "KEGG_CHRONIC_MYELOID_LEUKEMIA                                         True  \n",
       "KEGG_ACUTE_MYELOID_LEUKEMIA                                           True  \n",
       "KEGG_NON_SMALL_CELL_LUNG_CANCER                                       True  \n",
       "KEGG_AUTOIMMUNE_THYROID_DISEASE                                       True  \n",
       "KEGG_ALLOGRAFT_REJECTION                                              True  \n",
       "KEGG_GRAFT_VERSUS_HOST_DISEASE                                        True  \n",
       "KEGG_PRIMARY_IMMUNODEFICIENCY                                         True  \n",
       "KEGG_VIRAL_MYOCARDITIS                                                True  \n",
       "\n",
       "[84 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_p_values = emperic_p_values(expression, kegg_gmt, kegg_results, 100)\n",
    "kegg_p_values[kegg_p_values.Significantly_enriched == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gsea",
   "language": "python",
   "name": "gsea"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
